governance/ApiDeprecation.md

```diff
@@ -0,0 +1,116 @@
+1. _MUCH LATER_ (after the second issue is Closed, i.e. the second Release has actually happened)
+    1. Remove the code and create a PR
+    1. Set final issue to be Closed (will be closed by release team)
+    1. You're done!
+
+### Tracking (Informational)
+
+1. When a Release is imminent, sig-documentaion will (rip)grep the repo for all occurrences of `O3DE_DEPRECATION_NOTICE`
+1. They will confirm what stage the issue is in (following the issue chain) and add the info in the Description to the Release Notes.
+1. sig-documentation will then update the issue to Closed
+    1. When this happens, the engineer who began the deprecation will get a notification from GitHub of the change, and at this time they know it is okay to move to the next stage (add `AZ_DEPRECATED` (part 2), delete code (part 3)).
+
+### Find all O3DE_DEPRECATION_NOTICE tags
+
+```bash
+# prerequisites
+# ripgrep - https://github.com/BurntSushi/ripgrep/releases
+# cmder - https://cmder.net/
+ 
+rg O3DE_DEPRECATION_NOTICE\(GHI\-([0-9]+)\) --replace https://github.com/o3de/o3de/issues/$1 --iglob *.{h,hpp,c,cpp,inl,hxx} -o -I | sort | uniq > deprecations.txt
+ 
+# this will generate a list of Git Hub issue numbers to be updated when a Release is imminent.
+```
+
+### Advice
+
+Use of the `AZ_DEPRECATED` macro is now not allowed when first deprecating an API.
+
+It is now required to first use the comment identifier `O3DE_DEPRECATION_NOTICE`, and a second pass (corresponding to the second Git Hub issue), to add `AZ_DEPRECATED`.
+
+`AZ_DEPRECATED` cannot be used effectively as a first step in deprecation because the warning generated will cause a compiler error for many customers building with warnings as errors.
+
+The macro is useful for providing more information about how to update a particular call if it is being replaced.
+
+### FAQ
+
+#### Introducing alternative calls
+
+When first deprecating a function, it is advisable to not fix all call sites using the old version of the API. It may be wise to keep both variations active during the first stage of deprecation to ensure both calls do what you expect (e.g. Have the old call delegate to the new API and keep using that version for the first release, before the `AZ_DEPRECATED` macro is added). Depending on the situation it may be advisable to add a unit test to verify the behaviour of both calls. This helps avoid accidental, silent failures earlier.
+
+#### Deprecating EBus Functions
+
+When marking an EBus interface function `AZ_DEPRECATED`, any class that derives from that interface will generate `C4996 deprecation warnings`. This is caused by the `EBus::Handler` (the type that is actually inherited from) calling those functions internally in the EBus template code.
+
+To work around this, you need to wrap any class implementing that interface in
```

Comment:
Bummer? Is there some hackery we could implement to work around this, perhaps using an AZ_DEPRECATED_EBUS macro specifically for the job? Or, alternately could we provide a default implementation for the deprecated EBus function in the EBus interface class and remove the function from all the handlers?

Reply:
Hmm I'm not sure personally, it's possible @lumberyard-employee-dm might know 🤔 but I don't think there's an easy workaround unfortunately 😞

Reply:
Actually the AZ_DEPRECATED macro is pretty useless.
It gets into the whole situation that in order for a deprecation notice to be useful, it must be noticed by the user AND not cause a hard error while in the deprecation phase.

Those two conditions are at sort of at odds with each other.

Because we have warnings as errors on, the AZ_DEPRECATED macro would cause a hard compile error, that the user would notice, but it fails to satisfy the condition that the deprecation doesn't cause a hard error during it's deprecation phase.

If we suppress the warning, the the likelyhood of user's noticing the deprecation is quite low, causing the first condition to fail.

Reply:
It's worth mentioning the user (when seeing the warning and a potential hard failure) can add a suppression (without needing to make large code changes) to continue working.

The other useful thing with the warning is in the description you can add info about what to do to fix it (e.g. "Now use `GetWorldFromLocal` instead of `GetWorldTM`") which is way better than just getting a compile error with the call missing. So I still think it's useful.

This part relating to EBus is so we can deprecate an API but not have the warning trigger by other buses implementing the interface in our code (due to template issues with EBuses) so this is a very specific workaround for that case.